### PR TITLE
[Merged by Bors] - refactor(LinearAlgebra/Matrix/SesquilinearForm): Sesquilinear Maps and Matrices

### PR DESCRIPTION
**Pull request by mans0954**
Some of the concepts in `LinearAlgebra/Matrix/SesquilinearForm` can be generalised from Sesquilinear Forms to Sesquilinear Maps with little or no impact on the rest of Mathlib. This PR makes those generalisations.

A number of results in the `ToMatrix` section relating composition of maps to matrix multiplication could also be generalised, but they require a more genralised notion of matrix multiplication than we currently have available, so they are not included in the scope of this PR. Similarly for most of the results in the `Det` section.

---
See also: #9312

Hopefully this will replace #8256, in part at least.

- [x] depends on: #9475
- [x] depends on: #13042

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by mans0954** [comment]
!bench

**Comment by mans0954** [comment]
File                                            | metric       | master          | change               | this branch
------------------------------------------------|--------------|-----------------|----------------------|----------------
Mathlib.LinearAlgebra.Matrix.SesquilinearForm	| instructions |	380.472B |	49.57B 13.029 % | 	430.042B

**Comment by mans0954** [comment]
Before:
```
cumulative profiling times:
        attribute application 39.8ms
        compilation 1.2s
        dsimp 0.453ms
        elaboration 2.2s
        interpretation 0.163ms
        linting 343ms
        parsing 26.6ms
        simp 3s
        tactic execution 1.23s
        type checking 1.21s
        typeclass inference 28.2s
```
After:
```
cumulative profiling times:
        attribute application 38ms
        compilation 989ms
        dsimp 0.63ms
        elaboration 2.54s
        interpretation 0.147ms
        linting 322ms
        parsing 24.5ms
        simp 2.89s
        tactic execution 1.37s
        type checking 1.15s
        typeclass inference 30.5s
```

**Comment by mans0954** [comment]
!bench

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:201]: Should this be an `op` smul so that it actually generalizes what was there before? Maybe with appropriate `IsCentralScalar` assumptions or something.

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:390]: same here

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:201]: Would this be using `GroupTheory/GroupAction/Opposite`?

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (eric-wieser) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: This seems pretty fishy to me. I think the correct generality is
```lean
variable [Semiring R₁] [Semiring R₁'] [Semiring R₂] [Semiring R₂'] [AddCommMonoid M]
variable [Module R₁' M] [Module R₂' M] [SMulCommClass R₂' R₁' M]
variable (σ₁ : R₁ →+* R₁') (σ₂ : R₂ →+* R₂')

variable [Fintype n] [Fintype m]

/-- The map from `Matrix n n R` to bilinear maps on `n → R`.

This is an auxiliary definition for the equivalence `Matrix.toLinearMap₂'`. -/
def Matrix.toLinearMap₂'Aux (f : Matrix n m M) : (n → R₁) →ₛₗ[σ₁] (m → R₂) →ₛₗ[σ₂] M :=
  -- porting note: we don't seem to have `∑ i j` as valid notation yet
  mk₂'ₛₗ σ₁ σ₂ (fun (v : n → R₁) (w : m → R₂) => ∑ i, ∑ j, σ₂ (w j) • σ₁ (v i) • f i j)
    (fun _ _ _ => by simp only [Pi.add_apply, map_add, smul_add, sum_add_distrib, add_smul])
    (fun c v w => by
      simp only [Pi.smul_apply, smul_sum, smul_eq_mul, σ₁.map_mul, ← smul_comm _ (σ₁ c),
        MulAction.mul_smul])
    (fun _ _ _ => by simp only [Pi.add_apply, map_add, add_smul, smul_add, sum_add_distrib])
    (fun _ v w => by
      simp only [Pi.smul_apply, smul_eq_mul, _root_.map_mul, MulAction.mul_smul, smul_sum])
```
as this should then work for `(f : Matrix n m ℍ) : (n → ℍ) →ₗ[ℍ] (m → ℍ) →ₛₗ[starRingEquiv] ℍ`

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: More precisely, that definition works with:
```
example {R : Type*} [Ring R] [StarRing R] (f : Matrix n m R) :
    (n → R) →ₗ[R] (m → Rᵐᵒᵖ) →ₗ[Rᵐᵒᵖ] R :=
  Matrix.toLinearMap₂'Aux _ _ f
example {R : Type*} [Ring R] [StarRing R] (f : Matrix n m R) :
    (n → R) →ₗ[R] (m → R) →ₛₗ[(starRingEquiv (R := R)).toRingHom] R :=
  Matrix.toLinearMap₂'Aux _ _ f
```

Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: @eric-wieser would you then have:
```
variable [AddCommMonoid M₁] [Module R₁ M₁] [AddCommMonoid M₂] [Module R₂ M₂] [AddCommMonoid N₂]
  [Module R N₂] [SMulCommClass R R N₂]

variable {σ₁ : R₁ →+* R} {σ₂ : R₂ →+* R}

/-- The linear map from sesquilinear maps to `Matrix n m R` given an `n`-indexed basis for `M₁`
and an `m`-indexed basis for `M₂`.

This is an auxiliary definition for the equivalence `Matrix.toLinearMapₛₗ₂'`. -/
def LinearMap.toMatrix₂Aux (b₁ : n → M₁) (b₂ : m → M₂) :
    (M₁ →ₛₗ[σ₁] M₂ →ₛₗ[σ₂] N₂) →ₗ[R] Matrix n m N₂ where
  toFun f := of fun i j => f (b₁ i) (b₂ j)
  map_add' _f _g := rfl
  map_smul' _f _g := rfl
```
etc?

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: What's the assumption on `R` in that example? I think probably you're missing another ring or two there; the fact that conversion to a matrix is itself a linear operation is independent of the linear operation itself, so I am suspicious of the `→ₗ[R]`.

Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: So, something like:
```
variable [Semiring R] [Semiring R₁] [Semiring S₁] [Semiring R₂] [Semiring S₂]

variable [AddCommMonoid M₁] [Module R₁ M₁] [AddCommMonoid M₂] [Module R₂ M₂] [AddCommMonoid N₂]
  [Module R N₂] [Module S₁ N₂] [Module S₂ N₂] [SMulCommClass S₁ R N₂] [SMulCommClass S₂ R N₂] 
  [SMulCommClass S₂ S₁ N₂] 

variable {σ₁ : R₁ →+* S₁} {σ₂ : R₂ →+* S₂}

/-- The linear map from sesquilinear maps to `Matrix n m N₂` given an `n`-indexed basis for `M₁`
and an `m`-indexed basis for `M₂`.

This is an auxiliary definition for the equivalence `Matrix.toLinearMapₛₗ₂'`. -/
def LinearMap.toMatrix₂Aux (b₁ : n → M₁) (b₂ : m → M₂) :
    (M₁ →ₛₗ[σ₁] M₂ →ₛₗ[σ₂] N₂) →ₗ[R] Matrix n m N₂ where
  toFun f := of fun i j => f (b₁ i) (b₂ j)
  map_add' _f _g := rfl
  map_smul' _f _g := rfl
```
?

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: Yes, that looks sensible

Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: I'd got it in my head that we needed to be wary of using `SMulCommClass` too much for performance reasons, but maybe I got hold of the wrong end of the stick?

Comment (eric-wieser) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: That's almost certainly true, but also we don't want to do it at the expense of generalizing in the wrong direction.

Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:63]: @eric-wieser I'm afraid I don't quite have the brain power to see what `LinearMap.toMatrix₂` should become? I'm thinking something like:
```
variable [CommSemiring R] [CommSemiring R₁] [CommSemiring R₂]

variable [AddCommMonoid M₁] [Module R₁ M₁] [AddCommMonoid M₂] [Module R₂ M₂] [AddCommMonoid N₂]
  [Module R N₂] [Module R₁ N₂] [Module R₂ N₂] [SMulCommClass R₂ R N₂] [SMulCommClass R₁ R N₂]
  [SMulCommClass R₂ R₁ N₂] --[IsCentralScalar R N₂]

variable [DecidableEq n] [Fintype n]

variable [DecidableEq m] [Fintype m]

variable (b₁ : Basis n R₁ M₁) (b₂ : Basis m R₂ M₂)

/-- `LinearMap.toMatrix₂ b₁ b₂` is the equivalence between `R`-bilinear maps on `M` and
`n`-by-`m` matrices with entries in `R`, if `b₁` and `b₂` are `R`-bases for `M₁` and `M₂`,
respectively. -/
noncomputable def LinearMap.toMatrix₂ : (M₁ →ₗ[R₁] M₂ →ₗ[R₂] N₂) ≃ₗ[R] Matrix n m N₂ := sorry
```
We have:
```
#check (b₂.equivFun.arrowCongr (M₂₁ := N₂) (M₂₂ := N₂) (LinearEquiv.refl R₂ N₂) : (M₂ →ₗ[R₂] N₂) ≃ₗ[R₂] (m → R₂) →ₗ[R₂] N₂)
```
but then I'm not sure what to do next assuming `R₁` and `R₂` are not equal?

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (mans0954) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:692]: It no longer appears possible to use the dot notation on `M` here. Is this fixable @eric-wieser ?

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:61]: I think the problem is that I don't see a useful way to generalise `LinearEquiv.arrowCongr`.

Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:61]: Maybe something like:
```
def arrowCongr' {R M₁ M₂ M₂₁ M₂₂ : Sort _} [CommSemiring R] [AddCommMonoid M₁] [AddCommMonoid M₂]
    [AddCommMonoid M₂₁] [AddCommMonoid M₂₂]
    [Module R₁ M₁] [Module R₁ M₂] [Module R₁ M₂₁] [Module R₁ M₂₂]
    [Module R M₂₁] [Module R M₂₂]
    [SMulCommClass R₁ R M₂₁]
    [SMulCommClass R₁ R M₂₂]
    (e₁ : M₁ ≃ₗ[R₁] M₂) (e₂ : M₂₁ ≃ₗ[R₁] M₂₂) (h : IsLinearMap R e₂): (M₁ →ₗ[R₁] M₂₁) ≃ₗ[R] M₂ →ₗ[R₁] M₂₂
```

Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:61]: This then appears to work:
```
noncomputable def LinearMap.toMatrix₂ : (M₁ →ₗ[R₁] M₂ →ₗ[R₂] N₂) ≃ₗ[R] Matrix n m N₂ :=
  (b₁.equivFun.arrowCongr' (b₂.equivFun.arrowCongr' (LinearEquiv.refl R₂ N₂)
    ({ map_add := fun _ ↦ congrFun rfl, map_smul := fun _ ↦ congrFun rfl })) (
      { map_add := fun x ↦ congrFun rfl, map_smul := fun c ↦ congrFun (by exact rfl) })).trans
    LinearMap.toMatrix₂'
```

`(h : IsLinearMap R e₂)` should probably just be the scalar multiplication

Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:61]: Okay, I think I've fixed this now. Thanks!

Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:59]: @eric-wieser do you have an alternative suggestion for how this should look?

**Comment by mans0954** [comment]
!bench

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:279]:
```suggestion
-- TODO: Introduce matrix multiplication by matrices of scalars
```

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:166]: It seems that because of the changes, then `R` argument is almost never inferred, which is why you need `(R := R)` all over the place. I would just make this argument explicit to this function. Then everywhere, just use `R` instead of `(R := R)`. This may improve the output of the benchmark too.

Note that this behavior is not unexpected: in the previous iteration, the `R` argument could be inferred frequently from the first argument to the linear map when it was applied to some element, but that is no longer possible because `R` no longer appears in the type of that argument (or in the output argument).

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (j-loreaux) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:178]: Same comment about the `R` argument being explicit here.

**Review by j-loreaux — COMMENTED** [review]
I think `R` should be explicit, but otherwise I think this is okay.

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (mans0954) [reply Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:183]: Presumably the same applies to `Matrix.toLinearMapₛₗ₂'` and `LinearMap.toMatrixₛₗ₂'`?

**Comment by mans0954** [comment]
!bench

**Comment by mans0954** [comment]
@j-loreaux unfortunately the benchmarks look slightly worse (I'm assuming a larger percentage in red is worse?)

**Comment by j-loreaux** [comment]
That's such a small change, and only in a single file, that it's not a big deal. It could even be noise.

**Comment by MichaelStollBayreuth** [comment]
<b>General information:</b>
```text
lint:                                                  +3.5755 * 10⁹ (+0.0445%)
build:                                                 +49.877 * 10⁹ (+0.0406%)
```


<b>7 files got slower by at least 10⁹ instructions:</b>
```text
Mathlib.LinearAlgebra.Matrix.SesquilinearForm:         +35.147 * 10⁹ (+31.4%)
Mathlib.LinearAlgebra.Matrix.BilinearForm:             +3.8428 * 10⁹ (+4.73%)
Mathlib.Combinatorics.SimpleGraph.LapMatrix:           +2.9559 * 10⁹ (+12.6%)
Mathlib.LinearAlgebra.Matrix.PosDef:                   +2.1526 * 10⁹ (+4.43%)
Mathlib.LinearAlgebra.QuadraticForm.Basic:             +1.6895 * 10⁹ (+0.926%)
Mathlib.Algebra.Lie.Classical:                         +1.3773 * 10⁹ (+3.76%)
Mathlib.CategoryTheory.Category.Cat.Limit:             +1.3142 * 10⁹ (+1.38%)
```


<b>2 files got slower by at least 10%:</b>
```text
Mathlib.LinearAlgebra.Matrix.SesquilinearForm:                        +31.4%
Mathlib.Combinatorics.SimpleGraph.LapMatrix:                          +12.6%
```


<b>No file got faster by at least 10⁹ instructions.</b>


<b>No file got faster by at least 10%.</b>

**Comment by j-loreaux** [comment]
Thanks @MichaelStollBayreuth I'll let someone else decide. But I'm okay with this. The other option would be to have these general versions, and then more specific ones which are `abbrev`s of these to help Lean with elaboration. I expect that would help, but I'm not positive without trying it.

maintainer merge

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:132]: I would still like to use `toLinearMap₂'_apply` to unfold the definition, instead of doing so implicitly.
```suggestion
    Matrix.toBilin' M x y = ∑ i, ∑ j, x i * M i j * y j :=
  (Matrix.toLinearMap₂'_apply _ _ _).trans
    (by simp only [smul_eq_mul, mul_assoc, mul_comm, mul_left_comm])
```

File: Mathlib/LinearAlgebra/Matrix/BilinearForm.lean
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/BilinearForm.lean:256]: Idem:
```suggestion
  (Matrix.toLinearMap₂_apply _ _ _ _ _).trans
    (by simp only [smul_eq_mul, mul_assoc, mul_comm, mul_left_comm])
```

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:79]: Let's indent this slightly differently so the structure is clearer.
```suggestion
  have : (∑ i', ∑ j', (if i = i' then (1 : S₁) else (0 : S₁)) •
        (if j = j' then (1 : S₂) else (0 : S₂)) • f i' j') =
      f i j := by
```

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:347]: The old proof seems fine:
```suggestion
  (LinearMap.toMatrix₂' R).injective (by simp)
```

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:400]: `by exact pf` usually does exactly the same as `pf` itself.
```suggestion
    Matrix.toLinearMap₂ b₁ b₂ M x y = ∑ i, ∑ j, b₁.repr x i • b₂.repr y j • M i j :=
  Finset.sum_congr rfl fun _ _ => Finset.sum_congr rfl fun _ _ =>
    smul_algebra_smul_comm ((RingHom.id R) ((Basis.equivFun b₁) x _))
    ((RingHom.id R) ((Basis.equivFun b₂) y _)) (M _ _)
```

File: Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean
Comment (Vierkantor) [inline Mathlib/LinearAlgebra/Matrix/SesquilinearForm.lean:506]:
```suggestion
  simp_rw [LinearMap.toMatrix₂_compl₁₂ b₁ b₂, toMatrix_toLin, transpose_transpose]
```

**Review by Vierkantor — APPROVED** [review]
I found a few stylistic issues. Overall I think the slower build time is worth it for the increased generality.

So looks good to me with these changes, thanks!

bors d+

**Pull request closed without merging**
